lib/carbon/database.py

```diff
@@ -123,7 +124,13 @@ def write(self, metric, datapoints):
       whisper.update_many(path, datapoints)
 
     def exists(self, metric):
-      return exists(self.getFilesystemPath(metric))
+      if exists(self.getFilesystemPath(metric)):
+        return True
+      # if we're using hashed filenames and a non-hashed file exists then move it to the new name
+      if self.tag_hash_filenames and exists(self.getFilesystemPath(metric, False)):
+        os.rename(self.getFilesystemPath(metric, False), self.getFilesystemPath(metric))
```

Comment:
it shouldn't rename same file in different threads, right? no race condition here?

Reply:
The write process runs in a single dedicated thread, the whole procedure where the writer first checks if the file exists and creates it if needed isn't thread safe today.